File: day6/ts/main.ts
//algebraic solution from reddit u/themanushiya
import * as fs from 'fs';

function line2nums(_line: string) {
    const splits: string[] = _line.split(':')[1].trim().split(' ');
    const numstr: string[] = splits.filter((x) => x != '');
    return numstr.map((x) => Number(x));
}

function line2num(_line: string) {
    const splits: string[] = _line.split(':')[1].trim().split(' ');
    const numstr: string[] = splits.filter((x) => x != '');
    return Number(numstr.reduce((x, y) => x + y, ""));
}

function part1(_lines: string[]) {
    const tnums: number[] = line2nums(_lines[0]);
    const dnums: number[] = line2nums(_lines[1]);

    let res = 1;
    for (let i in tnums) {
        const t: number = tnums[i]
        const d: number = dnums[i]
        let low: number = (t - Math.sqrt((t * t) - 4 * d)) / 2;
        low = low % 1 ? Math.ceil(low) : low + 1;
        let top: number = (t + Math.sqrt((t * t) - 4 * d)) / 2;
        top = top % 1 ? Math.floor(top) : top - 1;
        res *= top - low + 1;
    }
    return res
}

function part2(_lines: string[]) {
    const t: number = line2num(_lines[0]);
    const d: number = line2num(_lines[1]);
    let low: number = (t - Math.sqrt((t * t) - 4 * d)) / 2;
    low = low % 1 ? Math.ceil(low) : low + 1;
    let top: number = (t + Math.sqrt((t * t) - 4 * d)) / 2;
    top = top % 1 ? Math.floor(top) : top - 1;
    return top - low + 1;
}

function showResult(str: string, func: Function, arg: any) {
    const start: number = performance.now();
    const res: any = func(arg)
    const end: number = performance.now();
    console.log(str, res);
    console.log("time[us] = ", (end - start) * 1000);
    return res;
}

const lines: string[] = fs.readFileSync('../input.txt', 'utf8').split('\n');
showResult("part1 = ", part1, lines);
showResult("part2 = ", part2, lines);

File: day6/cs/main.cs
//algebraic solution from reddit u/themanushiya
using System.Diagnostics;

int[] lineToNums(string line)
{
    int[] nums = line.Substring(line.IndexOf(':') + 1)
        .Trim(' ')
        .Split(' ')
        .Where(x => x.Length > 0)
        .Select(x => int.Parse(x))
        .ToArray();
    return nums;
}

long lineToNum(string line)
{
    string numstr = line.Substring(line.IndexOf(':') + 1)
        .Trim(' ')
        .Split(' ')
        .Where(x => x.Length > 0)
        .Aggregate("", (x , y) => x + y);
    return long.Parse(numstr);
}

long calcLow(long t, long d)
{
    double val = (t - Math.Sqrt((t * t) - 4 * d)) / 2;
    return (val == (long)val) ? (long)val + 1 : (long)Math.Ceiling(val);
}

long calcHi(long t, long d)
{
    double val = (t + Math.Sqrt((t * t) - 4 * d)) / 2;
    return (val == (long)val) ? (long)val - 1 : (long)Math.Floor(val);
}

long part1(string[] lines)
{
    int[] times = lineToNums(lines[0]);
    int[] dists = lineToNums(lines[1]);

    long res = 1;
    for (int i = 0; i < times.Length; i++)
    {
        long low = calcLow(times[i], dists[i]);
        long hi = calcHi(times[i], dists[i]);
        res *= hi - low + 1;
    }
    return res;
}

long part2(string[] lines)
{
    long time = lineToNum(lines[0]);
    long dist = lineToNum(lines[1]);

    long low = calcLow(time, dist);
    long hi = calcHi(time, dist);
    return hi - low + 1;
}
void show<T1, T2>(string s, Func<T1, T2> func, T1 arg)
{
    Stopwatch timer = Stopwatch.StartNew();
    T2 res = func(arg);
    timer.Stop();
    Console.WriteLine(String.Format(s + "{0}", res));
    Console.WriteLine(String.Format("time[us] = {0}", timer.Elapsed.TotalMicroseconds));
}

string[] lines = File.ReadAllLines(@"..\..\..\..\input.txt");
show<string[], long>("part1 = ", part1, lines);
show<string[], long>("part2 = ", part2, lines);
